source/BonusList.cpp

```diff
@@ -46,36 +46,57 @@ namespace Duel6 {
     }
 
     void BonusList::addRandomBonus() {
-        const Level &level = world.getLevel();
         bool weapon = (Math::random(2) == 1);
-        Int32 x, y, attempts = 0;
-
-        do {
-            attempts++;
-            x = Math::random(level.getWidth());
-            y = Math::random(level.getHeight());
-        } while (!isValidPosition(x, y, weapon) && attempts <= MAX_BONUS_ATTEMPTS);
+        std::vector<ValidPosition> validPositionList;
+        findValidPositions(validPositionList, weapon);
 
-        if (attempts > MAX_BONUS_ATTEMPTS) {
+        if (validPositionList.empty()) {
             return;
         }
+        std::shuffle(validPositionList.begin(), validPositionList.end(), Math::randomEngine);
+
+        ValidPosition position = validPositionList[0];
+        Int32 x = position.first, y = position.second;
+
+        bool isOverLimit = (weapons.size() + bonuses.size()) >= (validPositionList.size() / 4);
 
         if (weapon) {
             Int32 bullets = Math::random(10) + 10;
             weapons.push_back(LyingWeapon(Weapon::getRandomEnabled(settings), bullets, Vector(x, y)));
+            if(isOverLimit) {
+                weapons.pop_front();
+            }
         } else {
             const BonusType *type = BonusType::ALL[Math::random(BonusType::ALL.size())];
             bool random = Math::random(RANDOM_BONUS_FREQUENCY) == 0;
             Int32 duration = type->isOneTime() ? 0 : 13 + Math::random(17);
             bonuses.push_back(
                     Bonus(type, duration, Vector(x + 0.2f, y + 0.2f), random ? 0 : type->getTextureIndex()));
+            if(isOverLimit) {
+                bonuses.pop_front();
+            }
+        }
+    }
+
+    void BonusList::findValidPositions(ValidPositionList &validPositions, bool weapon) {
+        const Level &level = world.getLevel();
+        for (Int32 y = 0; y < level.getHeight(); y++) {
+            for (Int32 x = 0; x < level.getWidth(); x++) {
+                if (isValidPosition(x, y, weapon)) {
+                    validPositions.push_back(std::make_pair(x, y));
+                }
+            }
         }
     }
 
     bool BonusList::isValidPosition(const Int32 x, const Int32 y, bool weapon) {
         const Level &level = world.getLevel();
 
-        if (level.isWall(x, y, true) || (weapon && !level.isWall(x, y - 1, true))) {
+        if(level.isRaisingWater() && level.getWaterLevel() >= y) {
```

Comment:
Tohle zabrani pridavani bonusu do vody, kdyz zacne voda stoupat. Dohromady a tim weapons.pop_front() a bonuses.pop_front() to asi zapricini vycisteni bonusu a zbrani z vody. Je tohle dobrej napad?

Reply:
Nevim, jestli je to dobry napad, ale je to zamer :). Novy bonusy pribyvaji jen nad hladinu => vic bonusu v centru deni.